crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,532 @@
+use kitsune2_api::{K2Error, K2Result};
+
+/// In-memory store for time slice hashes.
+///
+/// The inner store will look something like this:
+/// `[ Block_0_3, Single_3_ABC, Block_4_2, Single_6_DEF ]`
+///
+/// Consecutive sequences of empty hashes are compressed into a single block. A block
+/// contains at least 1 hash.
+///
+/// Gaps are not permitted, so the representation starts out as a single block that spans
+/// the entire range of possible slice ids. As hashes are inserted, the block is split around
+/// the inserted hash.
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone))]
+pub(super) struct TimeSliceHashStore {
+    pub(super) highest_stored_id: Option<u64>,
+    inner: Vec<SliceHash>,
+}
+
+impl Default for TimeSliceHashStore {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum SliceHash {
+    Single { id: u64, hash: bytes::Bytes },
+    Block { id: u64, count: u64 },
+}
+
+#[cfg(test)]
+impl SliceHash {
+    pub(crate) fn id(&self) -> u64 {
+        match self {
+            SliceHash::Single { id, .. } => *id,
+            SliceHash::Block { id, .. } => *id,
+        }
+    }
+
+    pub(crate) fn end(&self) -> u64 {
+        self.id()
+            + match self {
+                SliceHash::Single { .. } => 0,
+                SliceHash::Block { count, .. } => *count,
+            }
+    }
+}
+
+impl TimeSliceHashStore {
+    pub(super) fn new() -> Self {
+        Self {
+            highest_stored_id: None,
+            inner: vec![SliceHash::Block {
```

Comment:
shouldn't this start empty? And then when the first Single is inserted we would also insert a Block before it? Otherwise we have to alter the size of this Block when the first Single is inserted, right?

Reply:
nevermind i see the logic for this below